### PR TITLE
Speed up tensor.resize_(sizes) when tensor has correct size

### PR DESCRIPTION
**Pull request by zou3519**
While using gbenchmark, I found `tensor.resize_({0})` would take 300ns
if tensor already has the correct size. This is important for
`at::empty({0})` perf because `at::empty` always calls `resize_`, which
in turn is a important for JIT perf: the fusion compiler creates empty
tensors and then `resize_`s them to computed sizes. Most of the 300ns is
due to DeviceGuard (200ns)

Summary of findings:
- `at::empty({0}, cuda)`: 851ns
- `empty_tensor.resize({0})`: 308ns
- `DeviceGuard(tensor)`: ctor + dtor: 200ns (Going to look into this
  next because it impacts `resize_` perf).
- vdispatch overhead (`tensor.resize_()` vs
  `at::native::resize__cuda(tensor)`): ~10ns

This PR rips out the TH `resize_` implementation and adds it to ATen
with the following modifications:
- DeviceGuard used only after the same-size check.
- Same-size check rewritten for simplicity. The new check doesn't
affect perf.
- empty_cpu / empty_cuda avoid the dispatch overhead to
tensor.resize_.

Timing with this PR:
- `at::empty({0}, cuda)`: 363ns
- `empty_tensor.resize_({0})`: 17ns

Future:
- Investigate `resize_(sizes)` slowness when `tensor.sizes() != sizes`
- Should tell resize_as_ to use the new resize_ implementation...
(because resize_as_ is in TH, it is calling the old TH resize_)

### CODE REVIEW

**Comment by SsnL** [comment]
Nice. Not a blocker. But now you can call AT functions in TH I believe. :)

File: aten/src/ATen/native/Resize.cpp [hunk 1]
```diff
@@ -0,0 +1,43 @@
+#include "ATen/ATen.h"
+#include "TH/THTensor.hpp"
+
+namespace at { namespace native {
+
+Tensor& resize__cpu(Tensor& self, IntList size) {
```
Comment (ezyang) [inline aten/src/ATen/native/Resize.cpp]: The return of the dastardly double underscore. (I laughed when I found out our codegen handles both naming conventions.)

File: aten/src/ATen/native/Resize.cpp [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  auto* self_ = self.unsafeGetTensorImpl();
+
+  int totalSize = 1;
+  int nDimension = size.size();
+  if(nDimension != self_->dim()) {
+    self_->resize_dim(nDimension);
+  }
+
+  for (int d = nDimension - 1; d >= 0; d--) {
+    self_->set_size(d, size[d]);
```
Comment (ezyang) [inline aten/src/ATen/native/Resize.cpp]: Oooh, please don't do it this way. Look at the implementation:

```
  virtual void set_size(int64_t dim, int64_t new_size) {
    sizes_.at(dim) = new_size;
    refresh_numel();
    refresh_contiguous();
  }
```

You're doing way more work than you need to. Use `set_sizes_and_strides` instead.

**Review by ezyang — CHANGES_REQUESTED** [review]
Sorry, I know you're just trying to make this fast, but I'd like you to do a few more things:

1. Don't call `set_size` manually
2. Stack on top of https://github.com/pytorch/pytorch/pull/12845 so you can get contiguous strides allocation for free. (Or copy paste it in; it's an easy enough merge conflict to resolve)
3. Delete the old implementation, replacing it with a call to the `at::native::` function

Thanks!

File: aten/src/ATen/native/Resize.cpp (hunk 1 above)
Comment (zou3519) [reply aten/src/ATen/native/Resize.cpp]: As in, our codegen handles `resize__cpu` and `resize_cpu_`?

**Comment by zou3519** [comment]
@ezyang I've addressed points (1) and (2). The conclusion for (3) is that there's no way to call aten native functions from TH at the moment but if that is not too hard to add I can punt this PR until that happens.

File: aten/src/ATen/core/TensorImpl.h [hunk 3]
```diff
@@ -348,6 +348,40 @@ struct CAFFE2_API TensorImpl : public c10::intrusive_ptr_target {
...
+  // sizes/strides are in bounds for the storage that is allocated;
+  // this is the responsibility of the caller
+  void set_sizes_contiguous(at::IntList new_size) {
+    AT_ASSERT(!is_variable());
+    auto old_dim = sizes_.size();
+    auto new_dim = new_size.size();
+    sizes_ = new_size.vec();
+
+    // TODO(rzou): replace with update_to_contiguous_strides
+    // in #12845 when that lands.
```
Comment (ezyang) [inline aten/src/ATen/core/TensorImpl.h]: Man, I hope we land that PR soon :)

File: aten/src/ATen/native/Resize.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/Resize.cpp]: yep

**Review by ezyang — APPROVED** [review]
okey-dokey

**Comment by zou3519** [comment]
@ezyang I figured out a compromise for (3): I've replaced the old TH implementation with the new implementation in this PR. I modified `TensorImpl::set_sizes_and_strides` to match the old `THTensor_(resizeNd)` semantics -- could you take a look and see if this is okay?

**Review by ezyang — APPROVED** [review]
OK. The code is still looks a bit copy-pastey, but I understand if you want to unblock.

**Comment by bddppq** [comment]
This PR has broken PyTorch ROCm tests:
https://ci.pytorch.org/jenkins/job/pytorch-builds/job/py2-clang7-rocmdeb-ubuntu16.04-test/1261//console

cc @iotamudelta 

```
06:33:27 test_zeros_like (__main__.TestUncoalescedSparse) ... ok
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_add_dense_sparse_mismatch (__main__.TestCudaSparse)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "test_sparse.py", line 933, in test_add_dense_sparse_mismatch
06:33:28     test_shape([3, 4, 0], [1, 4], [4, 4, 4, 0], [3, 4, 4, 0])
06:33:28   File "test_sparse.py", line 923, in test_shape
06:33:28     x = torch.zeros(dense_size, dtype=self.value_dtype, device=self.device)
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_add_dense_sparse_mismatch (__main__.TestCudaUncoalescedSparse)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "test_sparse.py", line 933, in test_add_dense_sparse_mismatch
06:33:28     test_shape([3, 4, 0], [1, 4], [4, 4, 4, 0], [3, 4, 4, 0])
06:33:28   File "test_sparse.py", line 923, in test_shape
06:33:28     x = torch.zeros(dense_size, dtype=self.value_dtype, device=self.device)
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ======================================================================
06:33:28 ERROR: test_reduction_empty (test_torch.TestTorch)
06:33:28 ----------------------------------------------------------------------
06:33:28 Traceback (most recent call last):
06:33:28   File "/var/lib/jenkins/workspace/test/test_torch.py", line 1047, in test_reduction_empty
06:33:28     self.assertEqual(torch.empty((2, 0, 1), device=device), fn(x, dim=2, keepdim=True))
06:33:28   File "/var/lib/jenkins/workspace/test/test_torch.py", line 1024, in <lambda>
06:33:28     ('sum', lambda *args, **kwargs: torch.sum(*args, **kwargs), 0),
06:33:28 RuntimeError: cuda runtime error (1011) : hipErrorInvalidValue at /var/lib/jenkins/workspace/aten/src/THC/generic/THCTensorMath.cu:25
06:33:28 
06:33:28 ----------------------------------------------------------------------
```

**Comment by jithunnair-amd** [comment]
This looks like more cases of "hipMemset with size 0"

**Comment by iotamudelta** [comment]
so please skip for now, this will be fixed with `ROCm 1.9.2`

**Pull request closed without merging**